### ISSUE TITLE
<ComplexInput />

### ISSUE THREAD

**Issue by malerba118** [chakra-ui/chakra-ui#5770]
### Description

Some complex inputs are hard to get right. For example number inputs are strangely hard to implement. I think there's a layer of abstraction that can mitigate the pain when implementing complex inputs that expect rigid syntax (eg json, css units, etc).

### Problem Statement/Justification

As mentioned complex inputs can be hard to get right. Imagine an input field where the user can include a number with a variety of suffixes (10px, 10%, 10em, 10vw, etc). If the user types an invalid suffix the value can't be accepted verbatim.

### Proposed Solution or API

It is common practice for complex inputs to allow the user to type anything and then to process the value on blur, either accepting or rejecting it. If rejected, the value will typically revert to the last known valid value.

I think a `ComplexInput` component could be a good addition to chakra.

Strangely, i've never seen this level of abstraction in the wild, but i think it could make for a good building block for complex inputs that you might encounter in an app like figma (eg: number input, input with unit (%, em, rem, px), json input, etc).

[Here's a codesandbox with a loose implementation of the propose api](https://codesandbox.io/s/complexinput-k9otd?file=/src/NumberInput.tsx).

Basically the idea is to pass parse/stringify implementations as props to the component.

```jsx
 <ComplexInput
      stringify={stringify}
      parse={parse}
      value={value}
      onChange={onChange}
 />
```

`onBlur` the component will call parse and decide if the input is valid. 

If so, `onChange` is called with the valid, parsed input. 

However, if the user's input is invalid the component will revert to the last known valid value and `onChange` will not be called.

**Comment by malerba118** [comment]
I just thought about this following bit, so i'm not totally sure of its validity, but i think another prop could be something like `coerce`, which would be an optional function that first attempts to salvage an invalid value. If it can be salvaged, then great, the coerced value can be accepted and `onChange` can be invoked. Otherwise if the invalid value can't be salvaged by the coerce function or no function is provided then the fallback behavior would be to revert to the last known valid value.

**Comment by Patrick-Ullrich** [comment]
Between existing libraries for masked inputs and as you pointed out you can do the parse in the `onBlur` handler, what is the advantage of creating another component?

**Comment by malerba118** [comment]
I think basically the advantage is tucking all of the logic illustrated below into a common layer, cause otherwise it's logic the user will need to implement themself if they want to implement a figma-style input. Don't have too much experience with input masking libs, but i feel like that's a bit of a different pattern where you actually prevent the user from typing characters which are disallowed (as opposed to letting them type anything and then processing the input on blur). Masking works sometimes for things like phone numbers but i can't see it working for something like a json input. Also imo sometimes masking creates a frustrating user experience.

```tsx
interface ComplexInputProps<T> extends Omit<InputProps, "value" | "onChange"> {
  stringify: (val: T) => string;
  parse: (val: string) => T;
  value: T;
  onChange: (value: T) => void;
}

function ComplexInput<T>({
  stringify,
  parse,
  value,
  onChange,
  ...otherProps
}: ComplexInputProps<T>) {
  const [draft, setDraft] = useState<string>(stringify(value));

  useEffect(() => {
    setDraft(stringify(value));
  }, [value]);

  const process = () => {
    try {
      let parsed = parse(draft);
      try {
        onChange(parsed);
      } catch (err) {
        // something went wrong in change handler
        // but don't want it to reset draft.
        console.warn(err);
      }
    } catch (err) {
      setDraft(stringify(value));
    }
  };

  return (
    <Input
      {...otherProps}
      onBlur={() => process()}
      onKeyDown={(e) => {
        if (e.key === "Enter") {
          e.currentTarget.blur();
        }
        otherProps.onKeyDown?.(e);
      }}
      value={draft}
      onChange={(e) => {
        setDraft(e.target.value);
      }}
    />
  );
}
```

**Comment by segunadebayo** [comment]
Hi @malerba118,

I think this is a great idea and it's a good candidate for a recipe in the docs. I don't think it's something we should add to the core library itself.

Feel free to create a recipe and we can post it in the docs, or even add it directly to the input component docs as well.

**Comment by malerba118** [comment]
@segunadebayo alright sir, sounds good to me. An obscure enough use case that I think a recipe makes sense over bloating the lib. I’ll try to throw a guide together 👍

**Comment by segunadebayo** [comment]
Great. Thank you @malerba118.
